language/stdlib/modules/validator_config_2.move

```diff
@@ -0,0 +1,142 @@
+        let ValidatorConfigUpdateToken { addr: _, sequence_number: _ } = update_token;
+    }
+    
+    // Returns true if addr has a published ValidatorConfig::T resource
+    public fun has(addr: address): bool { exists<T>(addr) }
+
+    // Getters for Config
+    public fun get_config(addr: address): Config acquires T { *&borrow_global<T>(addr).config }
+    public fun get_consensus_pubkey(config_ref: &Config): vector<u8> { *&config_ref.consensus_pubkey }
+    public fun get_sequence_number(config_ref: &Config): u64 { *&config_ref.sequence_number }
+
+    // Register the transaction sender as a candidate validator by creating a ValidatorConfig
+    // resource under their account. Note that only one such resource can be instantiated under an account.
+    public fun register_candidate_validator(
+        consensus_pubkey: vector<u8>,) {
+
+        move_to_sender<T>(
+            T {
+                config: Config {
+                    consensus_pubkey: consensus_pubkey,
+                    sequence_number: 0,
+                },
+                managing_delegated: false,
+                delegated_address: 0x0
+            }
+        );
+    }
+
+    // If the sender of the transaction has a ValidatorConfig::T resource,
+    // this function will delegate management of the ValidatorConfig::T resource to a delegated_address.
+    public fun set_delegated_address(delegated_address: address) acquires T {
+        Transaction::assert(LibraAccount::exists(delegated_address), 5);
+        let t_ref = borrow_global_mut<T>(Transaction::sender());
+        *(&mut t_ref.managing_delegated) = true;
+        *(&mut t_ref.delegated_address) = delegated_address;
+    }
+
+    // If the sender of the transaction has a ValidatorConfig::T resource,
+    // this function will revoke management capabilities from the delegated_address account.
+    public fun remove_delegated_address() acquires T {
+        let t_ref = borrow_global_mut<T>(Transaction::sender());
+        *(&mut t_ref.managing_delegated) = false;
+        *(&mut t_ref.delegated_address) = 0x0;
+    }
+
+    // Getters for delegated information.
+    // Those getters are only callable by the validator's account owner,
+    // others should be oblivious to how this account is managed.
+    public fun is_delegated_address_set(): bool acquires T {
+        let t_ref = borrow_global<T>(Transaction::sender());
+        t_ref.managing_delegated
+    }
+    public fun get_delegated_address(): address acquires T {
+        let t_ref = borrow_global<T>(Transaction::sender());
+        t_ref.delegated_address
+    }
+    
+    // Rotate validator's config invoked by delegate account.
+    // Here validator_address - is the address of the account of the validator whose consensus_pubkey is going to be rotated.
+    public fun rotate_consensus_pubkey_by_delegate(validator_address: address, new_consensus_pubkey: vector<u8>, proof: vector<u8>): ValidatorConfigUpdateToken acquires T {
```

Comment:
It might be easier if we combine this and `rotate_consensus_pubkey` (that way we will only need one transaction script to invoke this). One way this could work (I think) is:

```
let authorized_address = if (t_ref.managed_delegated) { t_ref.delegated_address } else { validator_address };
Transaction::assert(authorized_address == Transaction::sender(), 1)
```

Reply:
Yeh, but to get `t_ref` we will need to pass in validator_address in both cases, whereas in the `rotate_consensus_pubkey` this is the tx's sender.... that's why different number of arguments, but may be that's ok as we simplify the code.